Full GPIO Functionality

nullket opened the pull request:
**What this PR does**
This PR adds the ability to *optionally* set the two GPIOs of the cameras. If not set, it defaults to `input` in order to no break something.

**What can the GPIOs be set to?**
- Input
- Output Low
- Output High
- PWM (including frequency and duty cycle)
- Trigger (input)

**What else**
I have added an additional launch file to show an alternative way to create master-slaves setup. In this setup one camera creates a pwm output to trigger itself and other cameras

@jmackay2 again, I will ask for your time to review this please. I can imagine, that not everybody has the right cable. Therefore, you might can just review the code and see if it compiles fine in your setup without testing the GPIO functionality in case you do not have the right cable?

jmackay2 commented on launch/master_slaves_pwm_rgb8.launch:
```diff
@@ -0,0 +1,167 @@
+<!-- In this example launch file one camera creates a pwm signal on GPIO 2 which can be used to
```
Did you test this example? I am pretty sure the `<?xml version="1.0"?>` needs to be the first line, so probably move that up.

jmackay2 commented on launch/master_slaves_pwm_rgb8.launch:
```diff
@@ -0,0 +1,167 @@
+
+     <node pkg="nodelet" type="nodelet" name="ueye_cam_nodelet_slave1"
+          args="load ueye_cam/ueye_cam_nodelet nodelet_manager">
+     <param name="camera_name" type="str" value="master" /> <!-- == namespace for topics and services -->
```
Since the camera name and topic are the same, ROS is going to throw an error about trying to advertise the same service multiple times.

jmackay2 commented on launch/master_slaves_pwm_rgb8.launch:
```diff
@@ -0,0 +1,167 @@
+          args="load ueye_cam/ueye_cam_nodelet nodelet_manager">
+     <param name="camera_name" type="str" value="master" /> <!-- == namespace for topics and services -->
+     <param name="camera_topic" type="str" value="image_raw" />
+     <param name="camera_id" type="int" value="0" /> <!-- 0 = any camera; 1+: camera ID -->
```
Shouldn't we be incrementing the `camera_id` here?

jmackay2 commented on launch/master_slaves_pwm_rgb8.launch:
```diff
@@ -0,0 +1,167 @@
+
+     <node pkg="nodelet" type="nodelet" name="ueye_cam_nodelet_slave2"
+          args="load ueye_cam/ueye_cam_nodelet nodelet_manager">
+     <param name="camera_name" type="str" value="master" /> <!-- == namespace for topics and services -->
```
Same issue with the camera name and topic being the same.

jmackay2 commented on launch/master_slaves_pwm_rgb8.launch:
```diff
@@ -0,0 +1,167 @@
+          args="load ueye_cam/ueye_cam_nodelet nodelet_manager">
+     <param name="camera_name" type="str" value="master" /> <!-- == namespace for topics and services -->
+     <param name="camera_topic" type="str" value="image_raw" />
+     <param name="camera_id" type="int" value="0" /> <!-- 0 = any camera; 1+: camera ID -->
```
Same comment about incrementing the camera_id.

jmackay2 reviewed (commented):
Sorry it took me so long to get around to this. I don't have the cables around to test this, but I think the code looks fine. My main concern here is the new launch file, which seems like there could be some copy-paste errors. If you want to make sure you can run that file and test that it works as expected, that would be nice.

nullket commented:
> Sorry it took me so long to get around to this. I don't have the cables around to test this, but I think the code looks fine. My main concern here is the new launch file, which seems like there could be some copy-paste errors. If you want to make sure you can run that file and test that it works as expected, that would be nice.

I am using a **similar** launch file like this in our application - which works perfectly fine. In order to be more compatible with most cameras I used my working and tested launch file, removed some settings, changed the naming to be close to the existing sync file and added the larger comment at the top. As I have multiple (mixed synced and non-synced) cams connected I did not test the launch file again. Thanks a lot for the very precise review!! I corrected the launch file and hardcoded my camera_ids (so I could test) - but this is the same with already existing master/slave [launch file](https://github.com/anqixu/ueye_cam/blob/master/launch/master_slaves_rgb8.launch)

**tl;dr:** I just fixed the file. Thanks for the review. Please merge the PR when you think it is fine now.
